Archive/Imo/Imo1982Q1.lean

```diff
@@ -0,0 +1,125 @@
+
+On the other hand, we establish that $f(1980) \geq 660$
+from $5 \cdot f(1980) + 33\cdot f(3) \leq f(5\cdot 1980 + 33\cdot 1) = f(9999) = 3333$.
+
+We then conclude $f(1980) = 660$ and then eventually determine that either
+$f(1982) = 660$ or $f(1982) = 661$.
+
+In the latter case we derive a contradiction, because if $f(1982) = 661$ then
+$3334 = 5\cdot f(1982) + 29\cdot f(3) + f(2) \leq f(5\cdot 1982 + 29\cdot 3 + 2) = f(9999) = 3333$.
+-/
+
+namespace Imo1982Q1
+
+structure IsGood (f : ℕ+ → ℕ) : Prop where
+  /-- Satisfies the functional relation-/
+  rel: ∀ m n : ℕ+, f (m + n) = f m + f n ∨ f (m + n) = f m + f n + 1
+  f₂ : f 2 = 0
+  hf₃ : 0 < f 3
+  f_9999 : f 9999 = 3333
+
+namespace IsGood
+
+variable {f : ℕ+ → ℕ} (hf : IsGood f)
+include hf
+
+lemma f₁ : f 1 = 0 := by
+  have h : f 2 = 2 * f 1 ∨ f 2 = 2 * f 1 + 1 := by rw [two_mul]; exact hf.rel 1 1
+  obtain h₁ | h₂ := hf.f₂ ▸ h
+  · rw [eq_comm, mul_eq_zero] at h₁
+    apply h₁.resolve_left
+    norm_num
+  · cases Nat.succ_ne_zero _ h₂.symm
+
+lemma f₃ : f 3 = 1 := by
+    have h : f 3 = f 2 + f 1 ∨ f 3  = f 2 + f 1 + 1 := by apply hf.rel 2 1
+    rw [hf.f₁, hf.f₂, add_zero, zero_add] at h
+    have not_left : ¬ f 3 = 0 := by apply ne_of_gt hf.hf₃
+    rw [or_iff_right (not_left)] at h
+    exact h
+
+lemma superhomogeneous {m n : ℕ+} : ↑n * f m ≤ f (n * m) := by
+  induction n using PNat.recOn
+  case p1 => simp
+  case hp n' ih =>
+    have rel : f (n' * m + m) = f (n' * m) + f (m) ∨
+               f (n' * m + m) = f (n' * m) + f (m) + 1 := hf.rel (n' * m) m
+    calc
+    ↑(n' + 1) * f m = ↑n' * f m + f m := by rw [PNat.add_coe, add_mul, PNat.val_ofNat, one_mul];
+    _ ≤ f (n' * m) + f m := add_le_add_right ih (f m)
+    _ ≤ f (n' * m + m) := by
+      rcases rel with ( hl | hr)
+      · rw [hl]
+      · rw [hr]; nth_rewrite 1 [← add_zero (f m), ← add_assoc]; apply add_le_add_left (by norm_num)
+    _ = f ((n' + 1) * m) := by congr; rw [add_mul, one_mul]
+
+lemma superadditive {m n : ℕ+} : f m + f n ≤ f (m + n) := by
+  have h := hf.rel m n
+  rcases h with ( hl | hr )
+  · rw [hl]
+  · rw [hr]; nth_rewrite 1 [← add_zero (f n), ← add_assoc]; apply add_le_add_right (by norm_num)
+
+lemma superlinear {a b c d : ℕ+} : a * f b + c * f d ≤ f (a * b + c * d) :=
+  (add_le_add hf.superhomogeneous hf.superhomogeneous).trans hf.superadditive
+
+lemma part_1 : 660 ≤ f (1980) := by
+  rw [← mul_one (660), ← PNat.val_ofNat 660, ← hf.f₃]; apply hf.superhomogeneous
+
+lemma part_2 : f 1980 ≤ 660 := by
+  have h : 5 * f 1980 + 33 * f 3 ≤ 5 * 660 + 33 := by
+    calc (5 : ℕ+) * f 1980 + (33 : ℕ+) * f 3 ≤  f (5 * 1980 + 33 * 3) := by apply hf.superlinear
+    _ = f 9999 := by rfl
+    _ = 5 * 660 + 33 := by rw [hf.f_9999]
+  rw [hf.f₃, mul_one] at h
+  -- from 5 * f 1980 + 33 ≤ 5 * 660 + 33 we show f 1980 ≤ 660
+  exact le_of_mul_le_mul_left (le_of_add_le_add_right h) (Nat.succ_pos 4)
+
+end IsGood
+
+lemma imo1982_q1 {f : ℕ+ → ℕ} (hf : IsGood f) : f 1982 = 660 := by
```

Comment:
This should be outside of the `Imo1982Q1` namespace, otherwise the full name will be `Imo1982Q1.imo1982_q1`.